<a href="https://colab.research.google.com/github/VioletKimble/AIM_AHEAD/blob/main/AIM_AHEAD_Notebook_2_Analysis_and_Transformation_of_Structured_Patient_Datasets.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<center><img src="https://www.dropbox.com/scl/fi/bvwvq5lo4s32o2benttys/header.png?rlkey=iglro5700vvl27390wuw2j3yn&st=sia7zi1u&raw=1"/></center>

# 📊 **Analysis and Transformation of Structured Patient Datasets**

## 🎯 **Learning Objectives**
After completing this module, you will be able to:
* Explain the various components of a *pandas* dataframe.
* Use the `read_csv()` function to create new dataframes from real-world patient data.
* Manipulate and transform clinical data with *pandas* index and dataframe management functions.
* Illustrate the appropriate use of basic statistical functions to summarize a structured electronic health records dataset.
* Use matplotlib functions to generate clear graphs and plots.

> <img align="left" width="100" src="https://www.dropbox.com/scl/fi/h05iw3a4kbjybfjvxltnq/pandas_logo.png?rlkey=seib4xrmh599586m2e8fj62ph&st=eyojcdmu&raw=1" /> In this activity, you will learn how to use the [pandas](https://pandas.pydata.org) library (whose name is derived from the term *panel data*) to load, analyze, and process high quality structured patient data for downstream clinical machine learning tasks.

> 📌 **Note:** The name *pandas* is derived from the term *panel data*, referring to longitudinal measurements over time from a single subject.

## 🔍 Data-Centric Artificial Intelligence
**Data** is the fuel that powers artificial intelligence (AI), or more precisely, **machine learning** (ML) - the field of AI focused on statistical algorithms and techniques that can learn useful patterns from data without being explicitly programmed. Without volumes of machine-readable data, AI systems would not be as advanced as they are today, and may still be limited to the fixed rule-based AI systems from the 1950's and 60's.

While algorithms tend to garner the most publicity, in many (or even most) cases it is improvements in the data being used to train these algorithms that often leads to better and more useful AI models. This is especially true for AI using real-world patient data, such as clinical datasets derived from electronic health record (EHR) systems, which can contain errors, outliers, missing information, and other irregularities resulting from the complexities of real-world clinical care.

Source data undergoes several steps before being used to train a machine learning model, (e.g., acquisition, exploration, analysis, transformation, preprocessing, feature engineering, structuring for specific algorithms, and more). It is often these steps that require the most time and effort on an AI project. **[Data-centric AI](https://datacentricai.org)** is a recent school of thought which advocates for standardized practices which prioritize data curation over algorithm tuning.

As the landscape of AI algorithms and models continues to mature, systematic clinical data engineering and quality assurance will likely yield higher-quality models that drive better patient and health system outcomes. It is essential for clinical AI researchers to possess the skills, tools, and domain knowledge necessary for a comprehensive understanding of the data used to build healthcare AI systems.

**🚀 Let's get started!**

## 📥 Importing Modules
Similar to before, we'll start by `import`ing the `pandas` module, a necessary first step before we can use any of its functions. While not the focus of this notebook, we'll also be using [NumPy](https://numpy.org), a widely used scientific computing library, to demonstrate specific functionality of Pandas.

> 📌 **Note:** Python allows you to assign an *alias* to a module that you are importing by using the `as` keyword (`import MODULE_NAME ALIAS_NAME`). This allows you to use your alias throughout your code, instead of the orignal module name. This is most often done for coding brevity, and aliases are often chosen to be succinct - two-letter aliases are common). However, in rare cases, aliasing can be necessary. For example, if you'd like to import two modules with the same name, you can assign one to an alias to avoid the overlap.

>💡**Tip:** While you are free to select your own alias for any imported Python module, several widely used modules have acheived an unnoficial community consensus alias that you will see time and time again, including both `pandas` (with alias `pd`) and `numpy` (with alias `np`).

Let's import `numpy` and `pandas` using their common aliases.

In [1]:
import numpy as np
import pandas as pd

## 🗂️ DataFrame

The *DataFrame* is the primary data structure in the `pandas` library.

#### ✅ **Dataframes..**
* allow fast, flexible, and efficient data manipulation.
* are the natural data structure for **tabular** data.
* are two-dimensional, containing both **rows** and **columns**.
* consist of one or more **Series** (column)

> 📌 **Note:** `pandas` is built on top of `numpy`, which effectively means that most of the methods defined for **arrays** (the primary `numpy` data structure) also apply to DataFrames and other `pandas` objects.

#### 🤔 **What makes DataFrames and pandas so useful?**
There are many reasons why pandas is used for data analysis and engineering in AI and data science workflows, including (but not limited to):

* A large catalogue of convenient data management functions for automating common tasks
* A powerful and efficient interface to access and manipulate individual records
* Robust handling of missing values
* Relational database operations between DataFrames

>💡**Tip:** In a Jupyter notebook, if the last line of code in a cell is a variable, executing the cell will display the variable's value. We will be using this functionality extensively throughout this notebook, instead of the slightly more cumbersome `print` statements we saw in previous activities. This is convenient due to the cell-based nature of Jupyter notebooks, but it will not work in a regular Python script.

## Tabular Datasets for Machine Learning
The most straightforward type of dataset used for machine learning is **structured** (or *tabular*). A structured dataset can be thought of as a two-dimensional table or spreadsheet. In other words, the rows and columns have clearly defined meaning.

The typical format of a structured ML dataset is likely already familiar to you. Given a collection of `N` repeated observations of `D` variables, the corresponding structured dataset would be represented as a table with `N` rows and `D` columns. Each column would denote a different variable or attribute (*features* in ML parlance). Each of the `N` rows would represent a different sample of those `D` columns.

Consider a hypothetical dataset in which four data elements were recorded from 100 patients admitted to a local hospital: `age`, `sex`, `service`, and `elixhauser_comorbidity_index`. The resulting dataset would have 100 rows and 4 columns.

We can also say that the above dataset has a *shape* of `100 x 3`, denoting 100 rows and 3 columns per row.

> 📌 **Note:** Shape is an important concept for data transformation and model design. It is typically seen in the context of **arrays**.

## ➡️ Lists, Arrays, and Series

#### Lists
The `list` is the built-in Python data structure to store an ordered sequence of multiple elements. Lists are defined using the square bracket symbols `[]`, with elements inside the list separated by commas.

Let's create a simple list below using a variable called `spo2`. We'll also display its value using `print` and by typing the variable name on the last line of a code cell.

In [2]:
spo2 = [88, 99, 88, 85, 100, 98] # This is a Python list because it is in square brackets [ ]. It contains 6 integers.
print(spo2)
spo2 # This line has no effect, since it's not the last line of the cell.
spo2 # This line will display the value of spo2, since it's the last line of the cell.

[88, 99, 88, 85, 100, 98]


[88, 99, 88, 85, 100, 98]

#### Arrays
An `array` (commonly implemented with the `numpy` library) is a **vectorized** extension of a `list` that enables the use of numerous convenient data processing modules, and accelerates computations using low-level compiled code.  

> 📌 **Note:** Our introductory training has intentionally omitted the concept of **conditional statements** (if, then, else) and **loops** (for, while) because they can (and typically should) often be omitted in favor of vectorized linear algebra operations (using `numpy` and/or `pandas`), which are standard practice and can be orders of magnitude more efficient.

Let's create and display an array using our original `spo2` data:

In [3]:
spo2_array = np.array(spo2)
print(spo2_array) # Numpy arrays print similarly to Python lists, but do not include commas.
print(type(spo2_array)) # Arrays in numpy have type "numpy.ndarray"
spo2_array

[ 88  99  88  85 100  98]
<class 'numpy.ndarray'>


array([ 88,  99,  88,  85, 100,  98])

Arrays can also be two-dimensional (shown below) or more. This is why the `type` of a NumPy array is `ndarray`, i.e., extending the notion of 2-D or 3-D to an arbitrary number of dimensions N-D).

In [4]:
three_by_three = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
three_by_three

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

Finally, we can check the shape of a NumPy array with the `shape` attribute.

> 📌 **Note:** The `shape` attribute of one-dimensional vectors (like `spo2_array`) will typically omit the implied dimension `1` by default.

In [5]:
print(spo2_array.shape)
print(three_by_three.shape)

(6,)
(3, 3)


While `lists` and `arrays` may seem similar, working with `numpy` arrays opens up a vast catalogue of useful data manipulation functionality. See the [Official NumPy Documentation](https://numpy.org/doc/stable/) for more details.

#### Series
The one-dimensional vector in Pandas is the `Series`. It is similar to a NumPy array, but offers various advantages and flexibility for data processing tasks.  In a two-dimensional tabular dataset stored in a DataFrame, each **column** is a different `Series`.

Let's create a Series from our existing NumPy array:

In [6]:
spo2_series = pd.Series(spo2_array)
print(type(spo2_series)) # The type of a Pandas Series is "pandas.core.series.Series"
spo2_series # Notice the "dtype: int64", which means this Series contains integers.

<class 'pandas.core.series.Series'>


0     88
1     99
2     88
3     85
4    100
5     98
dtype: int64

We can also create a Series from a list of strings.

In [7]:
risk = ['high', 'low', 'high', 'high', 'low', 'low'] # This is a Python list, containing 6 strings.
risk_series = pd.Series(risk)
risk_series # Notice the "dtype: object", which is the dtype assigned to strings (or Series containing more than one type)

0    high
1     low
2    high
3    high
4     low
5     low
dtype: object

In [8]:
risk_series = pd.Series(risk) # Create a series from the values in the risk variable
pd.Series(risk) # Display the 'risk' series

0    high
1     low
2    high
3    high
4     low
5     low
dtype: object

## 🧱 Creating a Dataframe

> 📌 **Note:** This section uses a Python dictionary (`dict`), which is not a focus of this lesson.
>
> Briefly, from the [official Python documentation](https://docs.python.org/3/tutorial/datastructures.html):
> * It is best to think of a dictionary as a set of `key`: `value` pairs.
> * The keys within a dictionary must be unique
> * A pair of braces creates an empty dictionary: `{}`.
> * Placing a comma-separated list of `key`:`value` pairs within the braces adds initial `key`:`value` pairs to the dictionary.


DataFrames can be created in a variety of ways. One way is to provide the data manually as a Python dictionary that maps column names (the dictionary *keys*) to the 1-D vectors that should be inserted in that particular column (provided via the dictionary's *values*). The values could be Series, NumPy arrays, or even standard Python lists.

Let's create a DataFrame below using `spo2_array` (an array) and `risk` (a list).

>💡**Tip:** Another popular convention is to give the name `df` to an arbitrary DataFrame object. However, in practice, it is recommended to give assign descriptive names, especially if you are working with more than one DataFrame at the same time (common in practice).

In [9]:
df = pd.DataFrame({'spo2': spo2_array, 'risk': risk}) # Create the DataFrame
print(type(df)) # The type is "pandas.core.frame.DataFrame"
print(df.shape) # Similar to arrays, DataFrames also have the shape attribute.
df # View the DataFrame

<class 'pandas.core.frame.DataFrame'>
(6, 2)


spo2  risk
0    88  high
1    99   low
2    88  high
3    85  high
4   100   low
5    98   low

🎉 We've just created a small DataFrame `df` that has 6 rows and 2 columns, indicated by a `shape` of `(6,2)`. As you can see, pandas has no problem with the columns being different data types.

Let's extend these ideas to create a larger DataFrame with more rows and columns. In this example, we'll use the `numpy` function `randint()` to fill our DataFrame with random integers.

In [10]:
# Make some random numbers for data.
# This makes an array of random integers from 1-10, then reshapes it into a 500x4 matrix (500 rows and 4 columns)
random_data = np.random.randint(1,10, size=2000).reshape(500, 4)

# Create some example column names
column_names = ['columnA', 'columnB', 'columnC', 'columnD']

# Create the DataFrame from the random data we generated.
# Since the input matrix has 500 rows and 4 columns, the DataFrame will have 500 rows and 4 columns
df = pd.DataFrame(data=random_data, columns=column_names)
df # View the dataframe

columnA  columnB  columnC  columnD
0          3        4        7        8
1          6        6        6        8
2          6        2        1        5
3          5        3        9        9
4          1        6        4        1
..       ...      ...      ...      ...
495        3        3        8        5
496        8        4        4        7
497        8        7        9        6
498        8        3        6        7
499        7        8        1        5

[500 rows x 4 columns]

🎉 We've created a new DataFrame `df` that has 500 rows and 4 columns. You might also think of `df` as containing a two-dimensional matrix (array) which has a **shape** of 500x4 (shape is typically expressed as `# rows` x `# columns`). Since we used the same name `df` as our previous 6x2 DataFrame, the older DataFrame has been overwritten and is lost.

> 📌 **Note:** Notice that only the first and last 5 rows of data were displayed in detail. When viewing a large DataFrame in a Jupyter notebook, data will often be truncated similar to the above (but rest assured, the data still exists, even if it is not displayed.)

## 💾 Creating a DataFrame from an existing data file

Instead of manually creating a DataFrame from individual values (which would be impossibly tedious for large-scale datasets), a typical workflow involves loading the contents of an external file directly into a new DataFrame object. The file format you are most likely to encounter for AI is the **comma-separated value** `(.csv)` file.

#### 📕 Loading a DataFrame
We'll use the pandas function `read_csv()` to create a DataFrame from an existing `.csv` file. The `read_csv()` function has many optimal parameters to specify the file-loading process - see its [official documentation page](https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html) for more details. The only **required** parameter to specify for `read_csv()` is a string denoting the location of the `.csv` file to be loaded. This could be a local file path (e.g., "/Users/Me/data.csv"), or even a URL to a file hosted online.

In the following sections, we'll explore several methods using a real-world dataset of more than 4,000 body temperature measurements recorded from 100 unique patients during the course of their hospitalization.

> 🔗 The source data for our example dataset comes from [*Kallfelz et al. (2021). MIMIC-IV demo data in the OMOP common data model*](https://physionet.org/content/mimic-iv-demo-omop/0.9/).

⚠️ **The function below loads our dataset. However, it is not required to understand in detail and is presented for future reference as learners become more familiar with standards and data transformation.**

#### 🏗️ Assembling the dataset using the OMOP CDM (Optional)
The `get_temperature_data()` function below will return a DataFrame for our demonstration purposes. An in-depth understanding of this data-loading function is **not required** for this introductory `pandas` tutorial.

To build our DataFrame, we pull data from the `MEASUREMENT` and `OBSERVATION` tables in the [Observational Medical Outcomes Partnership (OMOP) Common Data Model (CDM)](https://ohdsi.github.io/TheBookOfOhdsi/CommonDataModel.html).
* In the `MEASUREMENT` table, we only keep rows with `measurement_concept_id = 3020891`, which corresponds to `Body temperature` measurements.
* In the `OBSERVATION` table, we only keep rows with `observation_concept_id = 3024265`, which corresponds to `Body temperature measurement site`.
* After joining both tables on unique patient identifier (`person_id`), unique hospitalization identifier (`visit_occurrence_id`), and  timestamp (`observation_datetime` and `measurement_datetime`), we are able to see all the desired information in a single DataFrame.
* For demonstration clarity, we rename several columns before returning the final DataFrame.

>💡**Tip:** The OHDSI-developed [Athena](https://athena.ohdsi.org) tool can be helpful for idenetifying OMOP concept_ids and better understanding standard vocabularies.

In [11]:
# This function builds and returns a DataFrame of real-world patient data (body temperature measurements).
# At this stage, a comprehensive understanding of this function is NOT REQUIRED.
# (For more advanced users, a summary of its operation is provided in the previous section.)
def get_temperature_data():
    measurement_file_url = 'https://www.dropbox.com/scl/fi/ecsom3v2v0lz52sx6r82v/measurement.csv?rlkey=cwxb5uizs4e2gwcg3ys0x1yoi&st=zlelcfjq&raw=1'
    observation_file_url = 'https://www.dropbox.com/scl/fi/smjlo0mh8op0i8obloha1/observation.csv?rlkey=3voe9dz4rqf8mdfwtaoasnvhs&st=svtmd4vd&raw=1'

    obs = pd.read_csv(observation_file_url,
                      usecols=['person_id', 'visit_occurrence_id', 'observation_datetime', 'observation_concept_id', 'value_as_string'],
                      parse_dates=['observation_datetime'])
    obs = obs[obs['observation_concept_id'] == 3024265] # 3024265 is the standard concept_id for "Body temperature measurement site"
    obs = obs[['person_id', 'visit_occurrence_id', 'observation_datetime', 'value_as_string']]

    meas = pd.read_csv(measurement_file_url,
                       usecols=['person_id', 'visit_occurrence_id', 'measurement_datetime', 'measurement_concept_id', 'value_as_number', 'unit_source_value'],
                       parse_dates=['measurement_datetime'])
    meas = meas[meas['measurement_concept_id'] == 3020891] # 3020891 is the standard concept_id for "Body temperature"
    meas = meas[['person_id', 'visit_occurrence_id', 'measurement_datetime', 'value_as_number', 'unit_source_value']]

    df = meas.merge(obs,
                    left_on=['person_id', 'visit_occurrence_id', 'measurement_datetime'],
                    right_on=['person_id', 'visit_occurrence_id', 'observation_datetime'],
                    how='left')
    df = df[['person_id', 'visit_occurrence_id', 'measurement_datetime', 'value_as_string', 'value_as_number', 'unit_source_value']]
    df = df.sort_values(by=['person_id', 'visit_occurrence_id', 'measurement_datetime'])
    df = df.reset_index(drop=True)

    # Rename columns for clarity
    df = df.rename(columns={
        'person_id': 'patient_id',
        'measurement_datetime': 'time',
        'value_as_string': 'temp_source',
        'value_as_number': 'temp_value',
        'unit_source_value': 'temp_unit'
    })

    df['temp_unit'] = df['temp_unit'].replace({
        '°F': 'Fahrenheit',
        '°C': 'Celsius'
    })

    return df

To load our example dataset, we'll call the `get_temperature_data()` function (which `returns` a DataFrame) and assign it to the `df` variable.

In [12]:
df = get_temperature_data()
df

patient_id  visit_occurrence_id                time  \
0    -9066461348710750663 -7972824337100083284 2185-01-17 19:39:00   
1    -9066461348710750663 -7972824337100083284 2185-01-17 21:25:00   
2    -9066461348710750663 -7972824337100083284 2185-01-17 22:30:00   
3    -9066461348710750663 -7972824337100083284 2185-01-18 00:00:00   
4    -9066461348710750663 -7972824337100083284 2185-01-18 00:14:00   
...                   ...                  ...                 ...   
4052  8805478484003283429  4821424463988433938 2114-06-20 20:00:00   
4053  8805478484003283429  4821424463988433938 2114-06-21 00:00:00   
4054  8805478484003283429  4821424463988433938 2114-06-21 04:00:00   
4055  8805478484003283429  4821424463988433938 2114-06-21 08:00:00   
4056  8805478484003283429  4821424463988433938 2114-06-21 12:00:00   

     temp_source  temp_value   temp_unit  
0           Oral        97.9  Fahrenheit  
1            NaN        36.6         NaN  
2            NaN        36.6         NaN  
3           Oral        98.9  Fahrenheit  
4            NaN        37.2         NaN  
...          ...         ...         ...  
4052        Oral        98.7  Fahrenheit  
4053        Oral        98.1  Fahrenheit  
4054        Oral        98.4  Fahrenheit  
4055        Oral        98.6  Fahrenheit  
4056        Oral        98.3  Fahrenheit  

[4057 rows x 6 columns]

This example dataset contains 4,057 rows and 6 columns related to body temperature measurements for 100 unique patients. A brief description of each of the 6 columns in `df` is shown below:

| Column Name |  Description |
| --- | --- |
|  `patient_id` |   a unique patient identifier (one for each unique patient in the dataset) |
|  `visit_occurrence_id` | a unique visit identifier (one for each hospital encounter in the dataset)
|  `time` |   the date and time of temperature measurement (dates have been shifted for privacy) |
|  `temp_source` |   the site where temperature was measured (e.g., oral, axillary, rectal) |
|  `temp_value` |   the numerical temperature measurement (e.g., 97.9, 36.6) |
|  `temp_unit` |   the measurement units for the `temp_value` (e.g., Fahrenheit, Celsius) |

## 🔎 Selecting Specific Data
DataFrames can store a large amount of data. In many AI projects, your DataFrames might contain millions of rows and thousands of columns. However, at any given time we may only be interested in a particular subset of data, or perhaps we'd simply like a preview of one part of the DataFrame to get a better intuitive understanding of what we have.

In the below examples, we'll see a few ways to view parts of a DataFrame, including subsets of rows, columns, or both.

#### ⬇️ Selecting a subset of rows
Let's start by viewing only some of a DataFrame's **rows**, which we will accomplish using the functions `head()` and `tail()`.

> 📌 **Note:** Pandas is a flexible library that allows for multiple ways to achieve the same result. Many of the examples we demonstrate in this lesson can be done in other ways. We encourage interested learners to explore the [official pandas documentation](https://pandas.pydata.org/docs/) for more details.

**✏️ Exercise:** In the code cell below, use the `head()` function to view the first 5 rows of our `df` DataFrame.
> **✨ Hint:** Given a dataframe `X`, calling `X.head()` will show the first `5` rows of the DataFrame.  

In [14]:
df.head()

patient_id  visit_occurrence_id                time temp_source  \
0 -9066461348710750663 -7972824337100083284 2185-01-17 19:39:00        Oral   
1 -9066461348710750663 -7972824337100083284 2185-01-17 21:25:00         NaN   
2 -9066461348710750663 -7972824337100083284 2185-01-17 22:30:00         NaN   
3 -9066461348710750663 -7972824337100083284 2185-01-18 00:00:00        Oral   
4 -9066461348710750663 -7972824337100083284 2185-01-18 00:14:00         NaN   

   temp_value   temp_unit  
0        97.9  Fahrenheit  
1        36.6         NaN  
2        36.6         NaN  
3        98.9  Fahrenheit  
4        37.2         NaN

**✏️ Exercise:** In the code cell below, use `head()` to view the first **10** rows of `df`.
> **✨ Hint:** You can also provide an input argument `n` to the `head()` function to show the first `n` rows (instead of the default `5` rows).

In [15]:
df.head(10)

patient_id  visit_occurrence_id                time temp_source  \
0 -9066461348710750663 -7972824337100083284 2185-01-17 19:39:00        Oral   
1 -9066461348710750663 -7972824337100083284 2185-01-17 21:25:00         NaN   
2 -9066461348710750663 -7972824337100083284 2185-01-17 22:30:00         NaN   
3 -9066461348710750663 -7972824337100083284 2185-01-18 00:00:00        Oral   
4 -9066461348710750663 -7972824337100083284 2185-01-18 00:14:00         NaN   
5 -9066461348710750663 -7972824337100083284 2185-01-18 01:54:00         NaN   
6 -9066461348710750663 -7972824337100083284 2185-01-18 04:00:00        Oral   
7 -9066461348710750663 -7972824337100083284 2185-01-18 05:00:00        Oral   
8 -9066461348710750663 -7972824337100083284 2185-01-18 06:00:00        Oral   
9 -9066461348710750663 -7972824337100083284 2185-01-18 07:00:00        Oral   

   temp_value   temp_unit  
0        97.9  Fahrenheit  
1        36.6         NaN  
2        36.6         NaN  
3        98.9  Fahrenheit  
4        37.2         NaN  
5        37.2         NaN  
6       100.9  Fahrenheit  
7       102.6  Fahrenheit  
8       102.1  Fahrenheit  
9       102.8  Fahrenheit

**✏️ Exercise:** In the code cell below, use `tail()` to view the **last** 5 rows of our `df` DataFrame.
> **✨ Hint:** Given a dataframe `X`, calling `X.tail()` will show the **last** `5` rows of the DataFrame.  

In [16]:
df.tail()

patient_id  visit_occurrence_id                time  \
4052  8805478484003283429  4821424463988433938 2114-06-20 20:00:00   
4053  8805478484003283429  4821424463988433938 2114-06-21 00:00:00   
4054  8805478484003283429  4821424463988433938 2114-06-21 04:00:00   
4055  8805478484003283429  4821424463988433938 2114-06-21 08:00:00   
4056  8805478484003283429  4821424463988433938 2114-06-21 12:00:00   

     temp_source  temp_value   temp_unit  
4052        Oral        98.7  Fahrenheit  
4053        Oral        98.1  Fahrenheit  
4054        Oral        98.4  Fahrenheit  
4055        Oral        98.6  Fahrenheit  
4056        Oral        98.3  Fahrenheit

**✏️ Exercise:** In the code cell below, use `tail()` to view the **last** `10` rows of `df`.
> **✨ Hint:** You can also provide an input argument `n` to the `tail()` function to show the last `n` rows (instead of the default `5` rows).

In [17]:
df.tail(10)

patient_id  visit_occurrence_id                time  \
4047  8692405834444096922 -6990694520804042573 2167-11-09 17:07:00   
4048  8692405834444096922 -6990694520804042573 2167-11-09 20:00:00   
4049  8805478484003283429  4821424463988433938 2114-06-20 10:55:00   
4050  8805478484003283429  4821424463988433938 2114-06-20 13:00:00   
4051  8805478484003283429  4821424463988433938 2114-06-20 16:00:00   
4052  8805478484003283429  4821424463988433938 2114-06-20 20:00:00   
4053  8805478484003283429  4821424463988433938 2114-06-21 00:00:00   
4054  8805478484003283429  4821424463988433938 2114-06-21 04:00:00   
4055  8805478484003283429  4821424463988433938 2114-06-21 08:00:00   
4056  8805478484003283429  4821424463988433938 2114-06-21 12:00:00   

     temp_source  temp_value   temp_unit  
4047        Oral        99.0  Fahrenheit  
4048        Oral        98.4  Fahrenheit  
4049        Oral        98.1  Fahrenheit  
4050        Oral        97.7  Fahrenheit  
4051        Oral        98.7  Fahrenheit  
4052        Oral        98.7  Fahrenheit  
4053        Oral        98.1  Fahrenheit  
4054        Oral        98.4  Fahrenheit  
4055        Oral        98.6  Fahrenheit  
4056        Oral        98.3  Fahrenheit

#### ➡️ Selecting a subset of columns
We'll now see a few ways to view a subset of a DataFrame's **columns**.

For a given DataFrame `df` and desired column `col`, one way to view a single column is to call `df['col']`, where the column name is wrapped in quotation marks.

**✏️ Exercise:** In the code cell below, use the above method to preview the `temp_value` column of our `df` DataFrame.

In [19]:
df['temp_value']

0       97.9
1       36.6
2       36.6
3       98.9
4       37.2
        ... 
4052    98.7
4053    98.1
4054    98.4
4055    98.6
4056    98.3
Name: temp_value, Length: 4057, dtype: float64

We can also adapt this method to view multiple columns, where instead of passing in a single string column name (e.g., `df['C']`, we will pass in a **list** of string column names, e.g. `df[['A', 'B', 'C']]` (notice the double brackets).

**✏️ Exercise:** In the code cell below, use the above method to view the `temp_value` **and** `temp_unit` columns of our `df` DataFrame.

In [21]:
df[['temp_value','temp_unit']]

temp_value   temp_unit
0           97.9  Fahrenheit
1           36.6         NaN
2           36.6         NaN
3           98.9  Fahrenheit
4           37.2         NaN
...          ...         ...
4052        98.7  Fahrenheit
4053        98.1  Fahrenheit
4054        98.4  Fahrenheit
4055        98.6  Fahrenheit
4056        98.3  Fahrenheit

[4057 rows x 2 columns]

Because many pandas functions and operations return a DataFrame object, these methods can be combined. In general, pandas functions are executed sequentially. While this is a simple example, chains of powerful data transforms can be accomplished with relatively few lines of pandas code.

> 📌 **Note:** Given a source DataFrame, it is important to understand whether a function will modify the original DataFrame values and/or return a **modified copy** (without touching the original values themselves). Many pandas functions provide the `inplace` parameter to control this behavior.

Here's one example of chaining multiple operations, where in a single line of code, we can display the first two values of `time` and `temp_value`.

In [22]:
df[['time', 'temp_value']].head(2)

time  temp_value
0 2185-01-17 19:39:00        97.9
1 2185-01-17 21:25:00        36.6

**✏️ Exercise:** In the code cell below, try more combinations of the previous techniques for viewing data subsets.

In [23]:
df['temp_unit'].tail()

4052    Fahrenheit
4053    Fahrenheit
4054    Fahrenheit
4055    Fahrenheit
4056    Fahrenheit
Name: temp_unit, dtype: object

## 🧮 Mathematical Operations

Various calculations can be performed on an entire column (i.e., the `Series`), either using library functions or even built-in arithmetic operators (e.g., `+`, `-`, `*`, `/`).

Let's assign the first 5 temperature measurements to a new variable called `first_five_temp`. These measurements all happen to be from patient `-9066461348710750663` over the course of about 4 hours of a single hospital encounter. We'll then demonstrate simple calculations using built-in Python arithmetic operators.

In [24]:
first_five_temp = df['temp_value'].head() # Create a new variable with the first 5 temperature measurements.
first_five_temp

0    97.9
1    36.6
2    36.6
3    98.9
4    37.2
Name: temp_value, dtype: float64

In [25]:
first_five_temp + 99.9 # Add an arbitrary constant number to every value

0    197.8
1    136.5
2    136.5
3    198.8
4    137.1
Name: temp_value, dtype: float64

In [26]:
first_five_temp - 1 # Subtract an arbitrary constant number to every value

0    96.9
1    35.6
2    35.6
3    97.9
4    36.2
Name: temp_value, dtype: float64

In [27]:
first_five_temp / 2 # Divide every value by an arbitrary constant number

0    48.95
1    18.30
2    18.30
3    49.45
4    18.60
Name: temp_value, dtype: float64

## 🎨 Creating Additional Columns

One way to add a new column to an existing DataFrame is with the following syntax: `df['NEW_COLUMN_NAME'] = NEW_COLUMN_VALUES`. Extending the mathematical operations above, we can quickly create and populate new columns based on operations performed on its current columns.

When we define an equation as below, the equation will be applied to every **row** of the DataFrame. In this way, we can create new columns from different combinations of existing columns.

Let's use an equation to convert Fahrenheit temperature to Celsius, and assign these values to a new column called `temp_celsius_naive`.

> 📌 **Note:** The `temp_value` column contains temperature values that have been recorded using different units (specified in the `temp_unit` column): Fahrenheit, Celsius, or the `temp_unit` was missing for the given measurement. Since the below equation will be applied to **every** row, the equation assumes that every value in the `temp_value` column is already measured in `Fahrenheit` (an incorrect assumption). In practice, it is critical to have a firm grasp on the entire dataset before performing modifications, because just like this example, many times these are **logical errors** that do not generate any error messages. For now, we'll call the new column `temp_celsius_naive`.

In [28]:
df['temp_celsius_naive'] = (df['temp_value'] - 32) * (5/9) # Apply the F to C equation, store the result in a new column.
df.head() # The df dataframe now has a new column.

patient_id  visit_occurrence_id                time temp_source  \
0 -9066461348710750663 -7972824337100083284 2185-01-17 19:39:00        Oral   
1 -9066461348710750663 -7972824337100083284 2185-01-17 21:25:00         NaN   
2 -9066461348710750663 -7972824337100083284 2185-01-17 22:30:00         NaN   
3 -9066461348710750663 -7972824337100083284 2185-01-18 00:00:00        Oral   
4 -9066461348710750663 -7972824337100083284 2185-01-18 00:14:00         NaN   

   temp_value   temp_unit  temp_celsius_naive  
0        97.9  Fahrenheit           36.611111  
1        36.6         NaN            2.555556  
2        36.6         NaN            2.555556  
3        98.9  Fahrenheit           37.166667  
4        37.2         NaN            2.888889

## 🗑️ Removing columns

We can also **drop** columns (i.e., delete them and all of their values from the DataFrame), using the syntax `df.drop(columns=LIST_OF_COLUMN_NAMES_TO_DROP)`.

Let's assume we discovered our logical error with the temperature conversion in the previous step, so now we'd like to remove the `temp_celsius_naive` column that we previously created.

> 🤔 **Question:** The code below applies the `drop` syntax correctly. Do you think the `temp_celsius_naive` column be removed?

In [29]:
df.drop(columns=['temp_celsius_naive']) # Drop the `temp_celsius_naive` column
df.head() # Checking to see if the column was removed

patient_id  visit_occurrence_id                time temp_source  \
0 -9066461348710750663 -7972824337100083284 2185-01-17 19:39:00        Oral   
1 -9066461348710750663 -7972824337100083284 2185-01-17 21:25:00         NaN   
2 -9066461348710750663 -7972824337100083284 2185-01-17 22:30:00         NaN   
3 -9066461348710750663 -7972824337100083284 2185-01-18 00:00:00        Oral   
4 -9066461348710750663 -7972824337100083284 2185-01-18 00:14:00         NaN   

   temp_value   temp_unit  temp_celsius_naive  
0        97.9  Fahrenheit           36.611111  
1        36.6         NaN            2.555556  
2        36.6         NaN            2.555556  
3        98.9  Fahrenheit           37.166667  
4        37.2         NaN            2.888889

🚫 The answer is **NO**, the `temp_celsius` column is still in our DataFrame!
* As mentioned earlier in this lesson, most functions do not (by default) modify the original input DataFrame.
* Instead, they perform the operation(s) and return a new, modified DataFrame object.
* This is the default behavior of functions that we saw in a previous lesson: if you do not assign this modified DataFrame to a new variable, it will be lost.

✅ Potential solutions:
1. Assign the result to a new variable: `df2 = df.drop(columns=['temp_celsius_naive'])`
2. Assign the result to the **same** variable name: `df = df.drop(columns=['temp_celsius_naive'])`, which will overwrite the original dataframe with the one that has been modified.
3. Use the `inplace=True` argument, available for many Pandas functions. After calling `df.drop(columns=['temp_celsius_naive'], inplace=True)`, the original DataFrame will be modified.

In the below code, we'll go with option 1: creating a new DataFrame `df2`.

In [30]:
df2 = df.drop(columns=['temp_celsius_naive'])
df2.head() # The column has been dropped, and does not exist in df2.

patient_id  visit_occurrence_id                time temp_source  \
0 -9066461348710750663 -7972824337100083284 2185-01-17 19:39:00        Oral   
1 -9066461348710750663 -7972824337100083284 2185-01-17 21:25:00         NaN   
2 -9066461348710750663 -7972824337100083284 2185-01-17 22:30:00         NaN   
3 -9066461348710750663 -7972824337100083284 2185-01-18 00:00:00        Oral   
4 -9066461348710750663 -7972824337100083284 2185-01-18 00:14:00         NaN   

   temp_value   temp_unit  
0        97.9  Fahrenheit  
1        36.6         NaN  
2        36.6         NaN  
3        98.9  Fahrenheit  
4        37.2         NaN

We can also check our original DataFrame to confirm that it still contains the `temp_celsius_naive` column:

In [31]:
df.head()

patient_id  visit_occurrence_id                time temp_source  \
0 -9066461348710750663 -7972824337100083284 2185-01-17 19:39:00        Oral   
1 -9066461348710750663 -7972824337100083284 2185-01-17 21:25:00         NaN   
2 -9066461348710750663 -7972824337100083284 2185-01-17 22:30:00         NaN   
3 -9066461348710750663 -7972824337100083284 2185-01-18 00:00:00        Oral   
4 -9066461348710750663 -7972824337100083284 2185-01-18 00:14:00         NaN   

   temp_value   temp_unit  temp_celsius_naive  
0        97.9  Fahrenheit           36.611111  
1        36.6         NaN            2.555556  
2        36.6         NaN            2.555556  
3        98.9  Fahrenheit           37.166667  
4        37.2         NaN            2.888889

**✏️ Exercise:** In the code cell below, create a new DataFrame called `df3` that contains everything in `df2` **except** for the `visit_occurrence_id` and `time` columns.
> **✨ Hint:** You want to drop the `visit_occurrence_id` and `time` columns.

In [41]:
df2.head()
df3 = df2.drop(columns=['visit_occurrence_id','time'])


## 🔭 Exploratory Data Analysis

Before diving into an AI project, taking the time to become intimately familiar with the data you will be modeling can save you time, energy, and headache later if things aren't turning out as you initially expected.  Exploratory data analysis can help you identify important data quality issues early in the AI process that can save you countless hours in the future.

#### 🤔 Why is exploratory analysis so important for any data science project?
* It can help identify major problems with the dataset that may challenge your assumptions and/or require returning to the data source or rethinking your entire question and/or approach.
* You can quickly identify data quality issues such as missing values, outliers, and otherwise incomplete, irrelevant, or misaligned data elements.
* A hands-on understanding of the data can influence your downstream selection of the best technical approach.
* A better understanding of your population is likely to lead to a more accurate interpretation of your results.
* Spending time early can help minimizing the amount of backtracking and repetition as issues are discovered later in pipeline.
* Furthermore, for deployed systems that rely on continuous integration (CI), deployment (CD), and training (CT) on new data over time, continuous iterations of rigorous data analysis can help spot issues such as data drift (where models trained on older data become less relevant for newer data streams)

Fortunately, pandas has a large collection of optimized and user-friendly methods for various types of data exploration and statistical analysis.

#### 📝 Summarizing your dataset

Similar to how we performed basic arithmetic with built-in Python operators like `+` and `-`, Pandas offers many one-line functions that compute a variety of statistics about your DataFrame:
* `sum` (sum the input values)
* `count` (count the number of input values), `value_counts` (count the number of each unique value)
* `min` (return the minimum value), `max` (return the maximum value)
* `mean` (return the mean value), `median` (median), `mode` (mode)
* `var` (variance), `std` (standard deviation)
* `quantile` (compute quantiles)

🔎 Let's take a look at a few examples using our example dataset of body temperature measurements.

First, let's compute the **mean** patient temperature measurement.

In [42]:
df['temp_value'].mean() # the mean() function will return a float (continuous number)

np.float64(88.19449839783091)

Hmm, something doesn't seem right. A mean body temperature of 88°F seems quite low (hypothermia occurs at 95°F). You may recall that we have both Celsius and Fahrenheit measurements in our dataset, which is skewing the results of our intended analysis.

> 🕔 We'll come back to this shortly (don't worry, we're eventually going to solve the problem!). For now, we'll continue with our demonstration of other useful pandas functions.

**✏️ Exercise:** In the code cell below, compute the **minimum** measured body temperature value.

In [43]:
df['temp_value'].min()

31.1

**✏️ Exercise:** In the code cell below, compute the **maximum** measured body temperature value.

In [44]:
df['temp_value'].max()

103.4

**✏️ Exercise:** CIn the code cell below, count the number of occurrences of **each unique** temperature measurement site (excluding missing values, which are represented in the DataFrame as `NaN`.
> **✨ Hint #1:** The `temp_source` column contains the temperature measurement site.

> **✨ Hint #2:** The `value_counts()` function can be used to calculate counts of each unique value (excluding missing values).

In [47]:
df['temp_source'].value_counts()

temp_source
Oral          2741
Blood          297
Axillary       288
Temporal       214
Rectal         122
Esophogeal      87
Tympanic        11
Name: count, dtype: int64

Pandas also offers `describe()`, a single function to compute multiple of the previous statistics across multiple columns at once.

In general, it's important to be aware of the values and data types contained in each column, as sometimes the statistics will not make sense. For example, `patient_id` is a randomized patient identifier string that carries no significance for arithmetic operations, so numerical calculations with these values will not be useful.

In [48]:
df.describe()

patient_id  visit_occurrence_id                           time  \
count  4.057000e+03         4.057000e+03                           4057   
mean  -7.797934e+16         5.376797e+17  2147-11-01 15:48:21.888095232   
min   -9.066461e+18        -9.149772e+18            2110-04-11 20:32:00   
25%   -4.353161e+18        -3.356179e+18            2131-05-23 17:00:00   
50%    5.792540e+17         1.482979e+18            2146-06-23 07:00:00   
75%    4.668337e+18         5.266191e+18            2161-11-19 12:16:00   
max    8.805478e+18         9.088822e+18            2201-12-13 18:00:00   
std    5.253521e+18         5.246692e+18                            NaN   

        temp_value  temp_celsius_naive  
count  4057.000000         4057.000000  
mean     88.194498           31.219166  
min      31.100000           -0.500000  
25%      97.500000           36.388889  
50%      98.200000           36.777778  
75%      98.900000           37.166667  
max     103.400000           39.666667  
std      23.127475           12.848597

## 🔪 Slicing Data with Boolean Expressions

☑️ Pandas provides multiple ways to select and process specific subsets (slices) of data based on flexible criteria.
* One way to accomplish this is by using **boolean expressions** that evaluate to either `True` or `False`.
* Boolean expressions often involve comparing values.
* For example, `x > 4` is a boolean expression that evaluates to `True` when a variable `x` is greater than 4, and `False` when `x` is less than or equal to 4.
* Another boolean expression might be `1 > 2`, which **always** evaluates to `False`.
* If we let `C` refer to our boolean expression (i.e., our selection criteria), we can use it to select a subset of our DataFrame with the syntax `df[C]`.
* Upon calling `df[C]`, the expression `C` will be evaluated for **every row** in the DataFrame, and only the rows where the expression evaluates to `True` will be returned in the result.
* You can think of this process as filtering the data based on a search criteria.

☑️ We can create selection criteria using familiar mathematical operators to compare values:
* `>`: greater than
* `<`: less than
* `>=`: greater than or equal to
* `<=`: less than or equal to
* `!=`: is not equal to
* `==`: is equal to a single value
* **Bonus:** `.isin(L)`: is equal to any value contained in the list `L`.

The code below will display all data where the patient's measured temperature value is greater than 103 degrees:

In [49]:
df[df['temp_value'] > 103] # Reminder: We're not assigning this result to a variable, so the below is not yet saved anywhere.

patient_id  visit_occurrence_id                time  \
2055   579254014084392336 -3332367749035126781 2117-03-18 18:00:00   
2063   579254014084392336 -3332367749035126781 2117-03-19 08:00:00   
3098  4668337230155062633  2572864492537913938 2116-12-07 13:00:00   
3100  4668337230155062633  2572864492537913938 2116-12-07 15:54:00   
3101  4668337230155062633  2572864492537913938 2116-12-07 18:11:00   
3103  4668337230155062633  2572864492537913938 2116-12-07 20:00:00   
3104  4668337230155062633  2572864492537913938 2116-12-07 21:00:00   

     temp_source  temp_value   temp_unit  temp_celsius_naive  
2055        Oral       103.3  Fahrenheit           39.611111  
2063        Oral       103.3  Fahrenheit           39.611111  
3098         NaN       103.3  Fahrenheit           39.611111  
3100        Oral       103.4  Fahrenheit           39.666667  
3101        Oral       103.4  Fahrenheit           39.666667  
3103        Oral       103.3  Fahrenheit           39.611111  
3104        Oral       103.4  Fahrenheit           39.666667

🧾 Let's break it down:
* In this example, the criteria `C` is `df['temp_value'] > 103`
* Inserting our criteria into `df[C]`, we have `df[df['temp_value'] > 103]`.
* This will evaluate the expression `df['temp_value'] > 103` (which evaluates to either `True` or `False` **for every row**, using that particular row's value of `temp_value`.
* Rows where the expression evaluates to `True` are included in the result, and the rest are excluded (but not dropped!)

**✏️ Exercise:** In the code cell below, display temperature measurements that are less than 60 degrees.
> **💡 Tip:** These measurements are likely to be using the `Celsius` unit, which could be one way to fill in missing values of the `temp_unit` column.

In [50]:
df['temp_source'].value_counts()

temp_source
Oral          2741
Blood          297
Axillary       288
Temporal       214
Rectal         122
Esophogeal      87
Tympanic        11
Name: count, dtype: int64

🎉 Let's now try using different criteria to select all data where the `temp_source` is `Axillary`.
* Recall that the symbol `==` will check for equality (we want the source to be equal to the single value `Core`).
* In this example, the criteria `C` is `df['temp_value'] == 'Axillary'`.
* Given the Pandas filtering syntax, what we end up with is `df[df['temp_source'] == 'Axillary']`.

In [51]:
df[df['temp_source'] == 'Axillary']

patient_id  visit_occurrence_id                time  \
63   -9066461348710750663 -7972824337100083284 2185-01-21 07:00:00   
64   -9066461348710750663 -7972824337100083284 2185-01-21 08:00:00   
65   -9066461348710750663 -7972824337100083284 2185-01-21 09:00:00   
67   -9066461348710750663 -7972824337100083284 2185-01-21 12:00:00   
163  -8659404739579738033   829078974010344533 2155-12-03 12:00:00   
...                   ...                  ...                 ...   
3987  8090044958540695372  2976307797036948160 2148-01-27 19:50:00   
3991  8090044958540695372  2976307797036948160 2148-01-28 16:00:00   
4012  8480470964666031560  4706395364776224406 2125-09-27 14:00:00   
4013  8480470964666031560  4706395364776224406 2125-09-27 14:10:00   
4014  8480470964666031560  4706395364776224406 2125-09-27 15:00:00   

     temp_source  temp_value   temp_unit  temp_celsius_naive  
63      Axillary        99.5  Fahrenheit           37.500000  
64      Axillary        99.5  Fahrenheit           37.500000  
65      Axillary        99.4  Fahrenheit           37.444444  
67      Axillary        98.8  Fahrenheit           37.111111  
163     Axillary        97.7  Fahrenheit           36.500000  
...          ...         ...         ...                 ...  
3987    Axillary        97.8  Fahrenheit           36.555556  
3991    Axillary        97.9  Fahrenheit           36.611111  
4012    Axillary        98.1  Fahrenheit           36.722222  
4013    Axillary        97.9  Fahrenheit           36.611111  
4014    Axillary        97.9  Fahrenheit           36.611111  

[288 rows x 7 columns]

**✏️ Exercise:** In the code cell below, select and display the subset of data in which the unit of temperature measurement is `Celsius`.

In [53]:
df[df['temp_unit'] == 'Celsius']

patient_id  visit_occurrence_id                time  \
12   -9066461348710750663 -7972824337100083284 2185-01-18 12:02:00   
13   -9066461348710750663 -7972824337100083284 2185-01-18 12:42:00   
14   -9066461348710750663 -7972824337100083284 2185-01-18 13:00:00   
16   -9066461348710750663 -7972824337100083284 2185-01-18 13:30:00   
17   -9066461348710750663 -7972824337100083284 2185-01-18 14:00:00   
...                   ...                  ...                 ...   
3932  8009761406928052418  5815945260189282379 2189-09-11 03:00:00   
3933  8009761406928052418  5815945260189282379 2189-09-11 04:00:00   
3934  8009761406928052418  5815945260189282379 2189-09-11 05:00:00   
3935  8009761406928052418  5815945260189282379 2189-09-11 06:00:00   
3936  8009761406928052418  5815945260189282379 2189-09-11 07:00:00   

     temp_source  temp_value temp_unit  temp_celsius_naive  
12        Rectal        39.4   Celsius            4.111111  
13           NaN        39.6   Celsius            4.222222  
14        Rectal        39.7   Celsius            4.277778  
16        Rectal        40.4   Celsius            4.666667  
17        Rectal        40.4   Celsius            4.666667  
...          ...         ...       ...                 ...  
3932       Blood        37.6   Celsius            3.111111  
3933       Blood        37.7   Celsius            3.166667  
3934       Blood        37.7   Celsius            3.166667  
3935       Blood        37.7   Celsius            3.166667  
3936       Blood        37.6   Celsius            3.111111  

[486 rows x 7 columns]

## ⛓️Combining multiple filtering criteria

Multiple criteria can be combined into larger boolean expressions using the `&` and`|` symbols. These more complex expressions still ultimately evaluate to either `True` or `False`.
* `C1 & C2`: This will select rows that match **both** criteria `C1` **and** `C2` (the overall expression is True only when both criteria are True.)
* `C1 & C2`: This will select rows that match **either** criteria `C1` **or** `C2` (When either one of the criteria is True (does not matter which), the overall expression is True).

Let's see how this technique works by searching for values between a desired range. If we'd like to select all data where the patient's measured temperature was between 99$^{\circ}$ and 100$^{\circ}$ (once again ignoring measurement units), we might do something like this:

In [54]:
df[(df['temp_value'] >= 99) & (df['temp_value'] <= 100)]

patient_id  visit_occurrence_id                time  \
59   -9066461348710750663 -7972824337100083284 2185-01-21 00:00:00   
63   -9066461348710750663 -7972824337100083284 2185-01-21 07:00:00   
64   -9066461348710750663 -7972824337100083284 2185-01-21 08:00:00   
65   -9066461348710750663 -7972824337100083284 2185-01-21 09:00:00   
69   -9066461348710750663 -7972824337100083284 2185-01-21 16:00:00   
...                   ...                  ...                 ...   
3970  8090044958540695372  2976307797036948160 2148-01-25 05:00:00   
4035  8527170356523164323  3284005162766861611 2180-07-23 20:00:00   
4043  8692405834444096922 -6990694520804042573 2167-11-09 00:00:00   
4044  8692405834444096922 -6990694520804042573 2167-11-09 04:00:00   
4047  8692405834444096922 -6990694520804042573 2167-11-09 17:07:00   

     temp_source  temp_value   temp_unit  temp_celsius_naive  
59          Oral        99.0     Celsius           37.222222  
63      Axillary        99.5  Fahrenheit           37.500000  
64      Axillary        99.5  Fahrenheit           37.500000  
65      Axillary        99.4  Fahrenheit           37.444444  
69          Oral        99.8  Fahrenheit           37.666667  
...          ...         ...         ...                 ...  
3970        Oral        99.2  Fahrenheit           37.333333  
4035        Oral        99.5  Fahrenheit           37.500000  
4043        Oral        99.3  Fahrenheit           37.388889  
4044        Oral        99.1  Fahrenheit           37.277778  
4047        Oral        99.0  Fahrenheit           37.222222  

[615 rows x 7 columns]

**✏️ Exercise:** In the code cell below, display all `Oral` temperature measurements that were greater than `103`$^{\circ}$.

In [55]:
df[(df['temp_source'] == 'Oral') & (df['temp_value'] > 103)]

patient_id  visit_occurrence_id                time  \
2055   579254014084392336 -3332367749035126781 2117-03-18 18:00:00   
2063   579254014084392336 -3332367749035126781 2117-03-19 08:00:00   
3100  4668337230155062633  2572864492537913938 2116-12-07 15:54:00   
3101  4668337230155062633  2572864492537913938 2116-12-07 18:11:00   
3103  4668337230155062633  2572864492537913938 2116-12-07 20:00:00   
3104  4668337230155062633  2572864492537913938 2116-12-07 21:00:00   

     temp_source  temp_value   temp_unit  temp_celsius_naive  
2055        Oral       103.3  Fahrenheit           39.611111  
2063        Oral       103.3  Fahrenheit           39.611111  
3100        Oral       103.4  Fahrenheit           39.666667  
3101        Oral       103.4  Fahrenheit           39.666667  
3103        Oral       103.3  Fahrenheit           39.611111  
3104        Oral       103.4  Fahrenheit           39.666667

In [56]:
df['time'].describe()

count                             4057
mean     2147-11-01 15:48:21.888095232
min                2110-04-11 20:32:00
25%                2131-05-23 17:00:00
50%                2146-06-23 07:00:00
75%                2161-11-19 12:16:00
max                2201-12-13 18:00:00
Name: time, dtype: object

In [57]:
df[df['temp_unit'] == 'Celsius'].describe()

patient_id  visit_occurrence_id                           time  \
count  4.860000e+02         4.860000e+02                            486   
mean  -6.254293e+16         1.684450e+18  2147-04-01 08:56:23.209876480   
min   -9.066461e+18        -7.972824e+18            2111-11-14 03:33:00   
25%   -4.353161e+18        -3.332368e+18  2125-02-27 16:03:44.999999488   
50%    5.792540e+17         3.186004e+18            2141-12-17 16:55:00   
75%    4.352191e+18         5.856429e+18            2185-01-18 15:30:00   
max    8.009761e+18         8.522635e+18            2189-09-11 07:00:00   
std    5.278618e+18         5.153880e+18                            NaN   

       temp_value  temp_celsius_naive  
count  486.000000          486.000000  
mean    37.215432            2.897462  
min     31.100000           -0.500000  
25%     36.300000            2.388889  
50%     36.950000            2.750000  
75%     37.500000            3.055556  
max     99.000000           37.222222  
std      4.946238            2.747910

## 🔬 Summarizing a Subset of Data
We can also apply the prior descriptive statistics functions on a specific subset of data. Let's combine slicing with the `describe()` function to analyze a few key statistics of all blood temperature measurements that were recorded in Celsius.

In [58]:
df[(df['temp_source'] == 'Blood') & (df['temp_unit'] == 'Celsius')].describe()

patient_id  visit_occurrence_id                           time  \
count  2.380000e+02         2.380000e+02                            238   
mean   9.888835e+17         4.644570e+18  2147-02-01 08:29:58.487394304   
min   -6.525153e+18        -4.740650e+18            2118-11-16 11:45:00   
25%   -4.873076e+18         3.186004e+18  2130-10-27 19:33:44.999999488   
50%    4.352191e+18         5.856429e+18            2140-03-25 21:30:00   
75%    6.543065e+18         7.387309e+18            2161-04-28 00:45:00   
max    8.009761e+18         8.522635e+18            2189-09-11 07:00:00   
std    5.791862e+18         3.947519e+18                            NaN   

       temp_value  temp_celsius_naive  
count  238.000000          238.000000  
mean    36.877731            2.709851  
min     34.600000            1.444444  
25%     36.500000            2.500000  
50%     36.900000            2.722222  
75%     37.375000            2.986111  
max     38.700000            3.722222  
std      0.660303            0.366835

We see from the above that there were 238 such measurements (`count=238`), the mean temperature was 36.88$^{\circ}$C, the median temperature was 36.9$^{\circ}$C, the minimum temperature was 36.5$^{\circ}$C, and the maximum temperature was 38.7$^{\circ}$C.

## 🧩 Putting it All Together

Let's return to our task of standardizing all temperature measurements to Celsius, where we can leverage several techniques discussed in this lesson.

* Earlier, we saw that we can't simply apply a formula to every `temp_value`, because some of these measurements are represented in Celsius and others in Fahrenheit.
* There's actually another issue: about 5% of our dataset is missing a value for `temp_unit`.
* Without knowing the measurement units, how can we be sure whether we should apply the temperature conversion formula (which would not be required if the original units are already in Celsius)?

**✏️ Exercise:** In the following code cell, determine how many `temp_unit` values are missing.
> **✨ Hint:** We want to count how many values in this column are `NaN` (instead of either `Fahrenheit` or `Celsius`). Counting values sounds like an opportunity to use the `value_counts()` function. By default, these counts will omit any `NaN` values, but if we supply the argument `dropna=True` to the function, it will include missing values (i.e., `NaN`) in the result.

> **💡 Tip:** Instead of returning integer counts, the `value_counts()` function will instead return relative fractions with the argument `normalize=True`.

**✏️ Exercise:** In the code cell below, analyze the distribution of measured values when the unit is missing. The `pd.isnull()` function can be used to evaluate whether a given value is missing (i.e., whether it is `NaN`.) the function as a boolean expression to select the rows that are missing a `temp_unit` value. What is the maximum temperature value in this context?
> **✨ Hint:** Consider the boolean expression `pd.isnull(df['temp_unit'])`.

Because the maximum temperature for measurements in this subset is well below even extreme Fahrenheit temperatures, we might decide that all such measurements were recorded in Celsius. Or, we might instead decide to leave it as missing (patterns of missingness can often be informative and may not simply be missing at random), remove those rows from the dataset altogether, or something else entirely.

> **📌 Note:** There is no "correct" solution! Data transformation decisions are often made based on domain knowledge, an understanding of the data-generating process, and the overarching goals of the analysis and AI.

For now, we'll go ahead and assume that the units were mistakenly omitted from the data, and should actually all be Celsius.

Let's go ahead and update the missing `temp_unit` values to be `Celsius`. There are multiple ways to accomplish this, but perhaps the most succinct is using the `fillna()` function, which replaces `NaN` values with the specified value. In the code below, we fill in missing values in the `temp_unit` column.

In [ ]:
df['temp_unit'] = df['temp_unit'].fillna('Celsius') # Remember, we must assign the result somewhere, or it has no effect.

**✏️ Exercise:** In the code cell below, determine how many `temp_unit` values are still missing from `df`.
> **✨ Hint:** There should be no more missing units.

Now that we've addressed the issue of missing temperature units, we still need to apply the temperature conversion formula correctly (i.e., only converting Fahrenheit measurements.) There are multiple ways to accomplish this.

**✏️ Exercise:** In the code cell below, create a new column in `df` called `temp_C` that contains every measured temperature represented in Celsius.

> **✨ Hint:** Memorizing the complete documentation and reference material for multiple Python libraries is not feasible (especially for those as large as pandas and for fast-moving fields such as AI. Part of learning how to code is learning how to seek out appropriate information that aids in solving a specific problem (this does not mean blindly copy/pasting code, which can be reckless and potentially dangerous).

> **✨ Hint:** If something isn't working as you expect, errors or warning messages can be extremely helpful in pointing you in the right direction. You will often find others in the online community who have already encountered and solved your problem (and if not, usually something similar).

> **✨ Hint:** There are multiple ways to approach this task. Depending on your approach, the highly rated answers to [this StackOverflow question](https://stackoverflow.com/questions/20625582/how-to-deal-with-settingwithcopywarning-in-pandas/53954986#53954986) may be helpful in learning about and addressing issues that we did not cover in this lesson.

**💪 More Practice** (Optional): Repeat the previous analyses of temperature measurements using the corrected DataFrame.

**🏆 This concludes our tutorial.** Please see the [Official Pandas User Guide](https://pandas.pydata.org/docs/user_guide/) for more examples and documentation for manipulating and analyzing data in Python.